libraries/chain/fork_database.cpp

```diff
@@ -603,15 +603,27 @@ namespace eosio::chain {
    }
 
    template<class BSP>
-   bool fork_database_t<BSP>::validated_block_exists(const block_id_type& id) const {
+   bool fork_database_t<BSP>::validated_block_exists(const block_id_type& id, uint32_t block_num) const {
       std::lock_guard g( my->mtx );
-      return my->validated_block_exists_impl(id);
+      return my->validated_block_exists_impl(id, block_num);
    }
 
+   // returns true if block `id`, or one of its ancestors >= block_num, is found in fork_db and `is_valid()`
+   // precondition: `id` is already in fork_db, so is a descendent of `root`
+   // ------------------------------------------------------------------------------------------------------
    template<class BSP>
-   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id) const {
-      auto itr = index.find( id );
-      return itr != index.end() && (*itr)->is_valid();
+   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id, uint32_t block_num) const {
+      assert(root && (root->id() == id || index.find(id) != index.end()));
```

Comment:
This seems overly restrictive. It may be true, but I don't see any reason why you need to enforce it. If `id` doesn't exist in the forkdb then the answer is false.

Reply:
https://github.com/AntelopeIO/spring/pull/788/commits/3a133d244e24c14f55b3b5ebdf4d936cabaf3b91